rust/private/module_extensions/toolchain.bzl

```diff
@@ -0,0 +1,77 @@
+"""Module extension for rust toolchain."""
+
+load("//rust:defs.bzl", "rust_common")
+load("//rust:repositories.bzl", "rust_register_toolchains", "rust_toolchain_tools_repository")
+load("//rust/platform:triple.bzl", "get_host_triple")
+load(
+    "//rust/private:repository_utils.bzl",
+    "DEFAULT_EXTRA_TARGET_TRIPLES",
+    "DEFAULT_NIGHTLY_VERSION",
+    "DEFAULT_STATIC_RUST_URL_TEMPLATES",
+)
+
+def _rust_impl(module_ctx):
+    # Allow the root module to define host tools. Otherwise, we'll fall back to
+    # the one defined in rules_rust.
+    for mod in module_ctx.modules:
+        if mod.tags.host_tools:
+            host_tools = mod.tags.host_tools[0]
+            host_triple = get_host_triple(module_ctx)
+
+            rust_toolchain_tools_repository(
+                name = "rust_host_tools",
+                exec_triple = host_triple.str,
+                target_triple = host_triple.str,
+                allocator_library = host_tools.allocator_library,
+                dev_components = host_tools.dev_components,
+                edition = host_tools.edition,
+                rustfmt_version = host_tools.rustfmt_version,
+                sha256s = host_tools.sha256s,
+                urls = host_tools.urls,
+                version = host_tools.version,
+            )
+            break
+
+    mod = module_ctx.modules[0]
+    for toolchain in mod.tags.toolchain:
+        rust_register_toolchains(
```

Comment:
The current naming for `rust_register_toolchains` is quite confusing and should probably be reworked separately.

If it's `rust_register_toolchains` it shouldn't have a `register_toolchain` bool flag. If it's just the configuration it should be called `rust_configure_toolchains` and the toolchain registration should be manually handled by users, even in the WORKSPACE case.